### PR TITLE
Fix Remote Recovery Being Retried for Removed Nodes

### LINKED ISSUE #57585: [CI] Failure in org.elasticsearch.cluster.MinimumMasterNodesIT
Failed on CI and once locally for me https://gradle-enterprise.elastic.co/s/3lz2nfmnfm5xm/

We're tripping an assertion in this test:

```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.cluster.MinimumMasterNodesIT" -Dtests.seed=2B9777EED69586A8 -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=Etc/UTC -Druntime.java=11
```

```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=310, name=elasticsearch[node_t2][scheduler][T#1], state=RUNNABLE, group=TGRP-MinimumMasterNodesIT]
Caused by: java.lang.AssertionError: Expected current thread [Thread[elasticsearch[node_t2][scheduler][T#1],5,TGRP-MinimumMasterNodesIT]] to not be the scheduler thread. Reason: [Blocking operation]
	at __randomizedtesting.SeedInfo.seed([2B9777EED69586A8]:0)
	at org.elasticsearch.threadpool.ThreadPool.assertNotScheduleThread(ThreadPool.java:731)
	at org.elasticsearch.common.util.concurrent.BaseFuture.blockingAllowed(BaseFuture.java:93)
	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:86)
	at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:56)
	at org.elasticsearch.indices.recovery.RecoverySourceHandler.lambda$acquireStore$22(RecoverySourceHandler.java:408)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:106)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:88)
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:36)
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:46)
	at org.elasticsearch.common.lease.Releasables.close(Releasables.java:51)
	at org.elasticsearch.common.lease.Releasables.lambda$releaseOnce$2(Releasables.java:105)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:106)
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:64)
	at org.elasticsearch.indices.recovery.RecoverySourceHandler.lambda$recoverToTarget$8(RecoverySourceHandler.java:248)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71)
	at org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:88)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:178)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListener(ListenableFuture.java:106)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.lambda$done$0(ListenableFuture.java:98)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:98)
	at org.elasticsearch.common.util.concurrent.BaseFuture.setException(BaseFuture.java:162)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.onFailure(ListenableFuture.java:135)
	at org.elasticsearch.action.StepListener.innerOnFailure(StepListener.java:67)
	at org.elasticsearch.action.NotifyOnceListener.onFailure(NotifyOnceListener.java:47)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71)
	at org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:88)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:178)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListener(ListenableFuture.java:106)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.lambda$done$0(ListenableFuture.java:98)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:98)
	at org.elasticsearch.common.util.concurrent.BaseFuture.setException(BaseFuture.java:162)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.onFailure(ListenableFuture.java:135)
	at org.elasticsearch.action.StepListener.innerOnFailure(StepListener.java:67)
	at org.elasticsearch.action.NotifyOnceListener.onFailure(NotifyOnceListener.java:47)
	at org.elasticsearch.indices.recovery.MultiFileTransfer.onCompleted(MultiFileTransfer.java:146)
	at org.elasticsearch.indices.recovery.MultiFileTransfer.handleItems(MultiFileTransfer.java:134)
	at org.elasticsearch.indices.recovery.MultiFileTransfer$1.write(MultiFileTransfer.java:79)
	at org.elasticsearch.common.util.concurrent.AsyncIOProcessor.processList(AsyncIOProcessor.java:108)
	at org.elasticsearch.common.util.concurrent.AsyncIOProcessor.drainAndProcessAndRelease(AsyncIOProcessor.java:96)
	at org.elasticsearch.common.util.concurrent.AsyncIOProcessor.put(AsyncIOProcessor.java:84)
	at org.elasticsearch.indices.recovery.MultiFileTransfer.addItem(MultiFileTransfer.java:90)
	at org.elasticsearch.indices.recovery.MultiFileTransfer.lambda$handleItems$4(MultiFileTransfer.java:126)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71)
	at org.elasticsearch.action.ActionListener$6.onFailure(ActionListener.java:292)
	at org.elasticsearch.action.ActionListener$4.onFailure(ActionListener.java:173)
	at org.elasticsearch.action.ActionListener$6.onFailure(ActionListener.java:292)
	at org.elasticsearch.action.support.RetryableAction$RetryingListener.onFailure(RetryableAction.java:149)
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:59)
	at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:545)
	at org.elasticsearch.indices.recovery.RemoteRecoveryTargetHandler$1.tryAction(RemoteRecoveryTargetHandler.java:258)
	at org.elasticsearch.action.support.RetryableAction$1.doRun(RetryableAction.java:98)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:691)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

The issue seems to be with the fact that we are scheduling some of the retrying in `org.elasticsearch.action.support.RetryableAction` on `SAME` so it runs on the scheduler thread. This in turn in some corner cases will use blocking `get` on a future and trip the above assertion.

### PR DESCRIPTION
**Pull request by original-brownbear**
If a node is disconnected we retry. It does not make sense
to retry the recovery if the node is removed from the cluster though.
=> added a check for the node being part of the cluster before retrying

Also, we were running the retry on the `SAME` pool which for each retry will
be the scheduler pool. Since the error path of the listener we use here
will do blocking operations when closing the resources used by the recovery
we can't use the `SAME` pool here since not all exceptions go to the `ActionListenerResponseHandler`
threading like e.g. `NodeNotConnectedException`.

Closes #57585

... non-issue label because this hasn't yet been released

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java [hunk 1]
```diff
@@ -261,7 +266,7 @@ public void tryAction(ActionListener<T> listener) {
 
             @Override
             public boolean shouldRetry(Exception e) {
-                return retriesSupported && retryableException(e);
+                return retriesSupported && clusterService.state().nodes().nodeExists(targetNode) && retryableException(e);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java:269]: This feels really dirty to me. I mainly opened this to illustrate where the problem lies, not sure this is an appropriate fix. Ideally, I'd hope we would have some mechanism for cancelling the retry if a node is dropped form the cluster but I couldn't find an easy way of building that (maybe someone else can?).

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java:269]: `ClusterStateObserver` is the magical thing here. It should be registered whenever we schedule a retry, and early cancel a scheduled retry when the target node drops from the cluster.

Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java:269]: Why don't we just have a single cluster state observer for `RemoteRecoveryTargetHandler`? We already have the whole cancellation infrastructure already built. Adding individual ones seems unnecessary.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java:269]: I didn't look at it in detail. Either is fine by me. Looks like @tbrooks8 volunteered to be a reviewer :)

**Comment by tbrooks8** [comment]
I’ll take a look at this tomorrow.

**Review by Tim-Brooks — CHANGES_REQUESTED** [review]
I feel like this can be done without creating these secondary codepaths (reaching back and forth between `PeerRecoverySourceService` and `RemoteRecoveryTargetHandler`). The original PR created a clear cancellation path. `PeerRecoverySourceService` listens for a shard closed event and then triggers `cancel` on existing `RemoteRecoveryTargetHandlers`. 

Can we not have a `Map<DiscoveryNode, Collection<RemoteRecoveryTargetHandler>>` in `OngoingRecoveries`? And it triggers the normal cancellations path:

```
synchronized void cancelOnNodeLeft(DiscoveryNode nodeThatLeft) {
            final Collection<RemoteRecoveryTargetHandler> handlers = nodeToHandlers.get(nodeThatLeft);
            if (handlers != null) {
                final List<Exception> failures = new ArrayList<>();
                for (RemoteRecoveryTargetHandler handler : handlers) {
                    handler.cancel();
                }
            }
        }
``` 

The normal `remove` call would cleanup the map. And then we don't have to do any of this logic where the `RemoteRecoveryTargetHandler` reaches up and messes with the state of `PeerRecoverySourceService`.

As an FYI, this path is slightly different than the cancellation path for the shard close. The shard close manually calls `close` on `RecoverySourceHandler` which triggers the cancel on cancellable threads before calling `cancel` on the `RecoveryTargetHandler`.

Calling `cancel` on just the `RecoveryTargetHandler` still will trigger the total cancellation because it will propagate an exception due to the requests failing. But I don't know if we want to unifying this cancellation path for both a node leaving and the shard closing? @ywelsch any thoughts? Probably does not need to be done in this PR. But I would like to avoid `RemoteRecoveryTargetHandler` having to reach up and mess with the state of `PeerRecoverySourceService`.

**Comment by original-brownbear** [comment]
@tbrooks8 thanks that sounds reasonable, let me try that.

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Tim!

**Pull request merged**